test/e2e/gateway_api_test.go

```diff
@@ -423,6 +428,284 @@ func testGatewayAPIRBAC(t *testing.T) {
+		}
+	})
+
+	if err != nil {
+		t.Fatalf("failed to create gateway with multiple listeners: %v", err)
+	}
+
+	gateway, err = assertGatewaySuccessful(t, operatorcontroller.DefaultOperandNamespace, "test-gateway-update")
+	if err != nil {
+		t.Fatalf("failed to create the test gateway: %v", err)
+	}
+
+	t.Logf("Created gateway %s with hostname %s", gateway.Name, "foo."+domain)
+
+	if err := assertExpectedDNSRecords(t, map[expectedDnsRecord]bool{
+		{dnsName: "foo." + domain + ".", gatewayName: "test-gateway-update"}: true,
+	}); err != nil {
+		t.Fatalf("DNSRecord expectations not met: %v", err)
+	}
+
+	// Fetch the latest version of the Gateway resource.
+	if err := kclient.Get(context.TODO(), types.NamespacedName{Namespace: operatorcontroller.DefaultOperandNamespace, Name: gateway.Name}, gateway); err != nil {
+		t.Fatalf("failed to get gateway resource: %v", err)
+	}
+
+	// Modify one gateway listener hostname
+	newHostname := gatewayapiv1.Hostname("baz." + domain)
+	gateway.Spec.Listeners[0].Hostname = &newHostname
+
+	if err := kclient.Update(context.TODO(), gateway); err != nil {
+		t.Fatalf("failed to update gateway %v: %v", gateway.Name, err)
+	}
+	t.Logf("Modified gateway %s's listener hostname from %s to: %s", gateway.Name, "foo."+domain+".", "baz."+domain+".")
+
+	if err := assertExpectedDNSRecords(t, map[expectedDnsRecord]bool{
+		{dnsName: "foo." + domain + ".", gatewayName: "test-gateway-update"}: false,
+		{dnsName: "bar." + domain + ".", gatewayName: "test-gateway-update"}: true,
+		{dnsName: "baz." + domain + ".", gatewayName: "test-gateway-update"}: true,
+	}); err != nil {
+		t.Fatalf("DNSRecord expectations not met: %v", err)
+	}
+
+	// Fetch the latest version of the Gateway resource.
+	if err := kclient.Get(context.TODO(), types.NamespacedName{Namespace: operatorcontroller.DefaultOperandNamespace, Name: gateway.Name}, gateway); err != nil {
+		t.Fatalf("failed to get gateway resource: %v", err)
+	}
+
+	// Delete one of the listeners
+	var updatedListeners []gatewayapiv1.Listener
+	for _, l := range gateway.Spec.Listeners {
+		if l.Name == "http-listener2" { //
+			continue
+		}
+		updatedListeners = append(updatedListeners, l)
+	}
+	gateway.Spec.Listeners = updatedListeners
+	if err := kclient.Update(context.TODO(), gateway); err != nil {
+		t.Fatalf("failed to update gateway with removed listener: %v", err)
+	}
+	t.Logf("Deleted listener %s from gateway %s.", "http-listener2", gateway.Name)
+
+	t.Logf("Checking that the dnsRecord for %s gets removed after removing the listener.", "bar."+domain+".")
+
+	if err := assertExpectedDNSRecords(t, map[expectedDnsRecord]bool{
+		{dnsName: "baz." + domain + ".", gatewayName: "test-gateway-update"}: true,
+		{dnsName: "bar." + domain + ".", gatewayName: "test-gateway-update"}: false,
+	}); err != nil {
+		t.Fatalf("DNSRecord expectations not met: %v", err)
+	}
+
+	if err := kclient.Delete(context.TODO(), gateway); err != nil {
+		if errors.IsNotFound(err) {
+			return
```

Comment:
Why just `return`? Your gateway should be there, if someone deleted it it's a test `Error()`.